File: src/backend/app/routers/swarm.py
"""Swarm playground routes: create/list/get runs, start/pause control, and
NDJSON live event streaming."""
from __future__ import annotations

import asyncio
import json
from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import StreamingResponse
from psycopg import Connection
from psycopg.types.json import Jsonb
from pydantic import BaseModel, Field

from app.auth import CurrentUser
from app.db import db_conn
from app.services import swarm as swarm_service
from app.services.swarm_broker import broker

router = APIRouter(prefix="/api/projects", tags=["swarm"])


# --------------------------------------------------------------------------- #
# Schemas
# --------------------------------------------------------------------------- #


SwarmType = Literal["aco", "pso", "abc", "firefly"]
RunModel = Literal["claude-opus-4-7", "claude-sonnet-4-6", "claude-haiku-4-5"]
ThoughtLevel = Literal["minimal", "standard", "deep", "extreme"]
RunStatus = Literal["idle", "running", "paused", "completed", "failed"]


class SwarmRunConfig(BaseModel):
    swarm_type: SwarmType = "aco"
    model: RunModel = "claude-sonnet-4-6"
    num_agents: int = Field(default=8, ge=2, le=64)
    thought_level: ThoughtLevel = "standard"
    randomness: float = Field(default=0.3, ge=0.0, le=1.0)
    pheromone_strength: float = Field(default=0.6, ge=0.0, le=1.0)


class SwarmRunSummary(BaseModel):
    id: UUID
    project_id: UUID
    config: SwarmRunConfig
    status: RunStatus
    current_turn: int
    best_score: float | None
    created_at: datetime
    updated_at: datetime


class SwarmRunDetail(SwarmRunSummary):
    best_attempt_id: UUID | None
    best_system_text: str | None
    best_user_template: str | None
    error: str | None


class SwarmAttempt(BaseModel):
    id: UUID
    turn: int
    agent_idx: int
    parent_attempt_id: UUID | None
    status: str
    score: float | None
    pheromone: float
    x: float | None
    y: float | None
    system_text: str | None
    user_template: str | None
    predicted_output: dict[str, Any] | None
    error: str | None


# --------------------------------------------------------------------------- #
# Auth helper
# --------------------------------------------------------------------------- #


def _ensure_owner(conn: Connection, project_id: UUID, user_id: str) -> None:
    with conn.cursor() as cur:
        cur.execute(
            "SELECT 1 FROM projects WHERE id = %s AND user_id = %s",
            (project_id, user_id),
        )
        if cur.fetchone() is None:
            raise HTTPException(status_code=404, detail="Project not found")


def _ensure_run_owner(
    conn: Connection, project_id: UUID, run_id: UUID, user_id: str
) -> None:
    _ensure_owner(conn, project_id, user_id)
    with conn.cursor() as cur:
        cur.execute(
            "SELECT 1 FROM swarm_runs WHERE id = %s AND project_id = %s",
            (run_id, project_id),
        )
        if cur.fetchone() is None:
            raise HTTPException(status_code=404, detail="Run not found")


# --------------------------------------------------------------------------- #
# REST
# --------------------------------------------------------------------------- #


@router.post(
    "/{project_id}/swarm-runs", response_model=SwarmRunSummary, status_code=201
)
def create_run(
    project_id: UUID,
    body: SwarmRunConfig,
    user_id: CurrentUser,
    conn: Connection = Depends(db_conn),
) -> SwarmRunSummary:
    _ensure_owner(conn, project_id, user_id)
    with conn.cursor() as cur:
        cur.execute(
            """
            INSERT INTO swarm_runs (project_id, config)
            VALUES (%s, %s)
            RETURNING id, project_id, config, status, current_turn,
                      best_score, created_at, updated_at
            """,
            (project_id, Jsonb(body.model_dump())),
        )
        row = cur.fetchone()
    conn.commit()
    return SwarmRunSummary(**row)


@router.get("/{project_id}/swarm-runs", response_model=list[SwarmRunSummary])
def list_runs(
    project_id: UUID,
    user_id: CurrentUser,
    conn: Connection = Depends(db_conn),
) -> list[SwarmRunSummary]:
    _ensure_owner(conn, project_id, user_id)
    with conn.cursor() as cur:
        cur.execute(
            """
            SELECT id, project_id, config, status, current_turn,
                   best_score, created_at, updated_at
            FROM swarm_runs
            WHERE project_id = %s
            ORDER BY created_at DESC
            """,
            (project_id,),
        )
        rows = cur.fetchall()
    return [SwarmRunSummary(**r) for r in rows]


@router.get(
    "/{project_id}/swarm-runs/{run_id}", response_model=SwarmRunDetail
)
def get_run(
    project_id: UUID,
    run_id: UUID,
    user_id: CurrentUser,
    conn: Connection = Depends(db_conn),
) -> SwarmRunDetail:
    _ensure_run_owner(conn, project_id, run_id, user_id)
    with conn.cursor() as cur:
        cur.execute(
            """
            SELECT r.id, r.project_id, r.config, r.status, r.current_turn,
                   r.best_attempt_id, r.best_score, r.error,
                   r.created_at, r.updated_at,
                   ba.system_text AS best_system_text,
                   ba.user_template AS best_user_template
            FROM swarm_runs r
            LEFT JOIN swarm_attempts ba ON ba.id = r.best_attempt_id
            WHERE r.id = %s
            """,
            (run_id,),
        )
        row = cur.fetchone()
    if row is None:
        raise HTTPException(status_code=404, detail="Run not found")
    return SwarmRunDetail(**row)


@router.get(
    "/{project_id}/swarm-runs/{run_id}/attempts",
    response_model=list[SwarmAttempt],
)
def list_attempts(
    project_id: UUID,
    run_id: UUID,
    user_id: CurrentUser,
    conn: Connection = Depends(db_conn),
) -> list[SwarmAttempt]:
    _ensure_run_owner(conn, project_id, run_id, user_id)
    with conn.cursor() as cur:
        cur.execute(
            """
            SELECT id, turn, agent_idx, parent_attempt_id, status,
                   score, pheromone, x, y,
                   system_text, user_template, predicted_output, error
            FROM swarm_attempts
            WHERE run_id = %s
            ORDER BY turn, agent_idx
            """,
            (run_id,),
        )
        rows = cur.fetchall()
    return [SwarmAttempt(**r) for r in rows]


@router.post("/{project_id}/swarm-runs/{run_id}/start")
def start_run(
    project_id: UUID,
    run_id: UUID,
    user_id: CurrentUser,
    conn: Connection = Depends(db_conn),
) -> dict[str, Any]:
    _ensure_run_owner(conn, project_id, run_id, user_id)
    started = swarm_service.start_run(run_id)
    return {"started": started, "status": swarm_service._get_run_status(run_id)}


@router.post("/{project_id}/swarm-runs/{run_id}/pause")
async def pause_run(
    project_id: UUID,
    run_id: UUID,
    user_id: CurrentUser,
    conn: Connection = Depends(db_conn),
) -> dict[str, Any]:
    _ensure_run_owner(conn, project_id, run_id, user_id)
    await swarm_service.request_pause(run_id)
    return {"status": swarm_service._get_run_status(run_id)}


@router.delete(
    "/{project_id}/swarm-runs/{run_id}",
    status_code=204,
    response_model=None,
)
def delete_run(
    project_id: UUID,
    run_id: UUID,
    user_id: CurrentUser,
    conn: Connection = Depends(db_conn),
) -> None:
    _ensure_run_owner(conn, project_id, run_id, user_id)
    with conn.cursor() as cur:
        cur.execute("DELETE FROM swarm_runs WHERE id = %s", (run_id,))
    conn.commit()


# --------------------------------------------------------------------------- #
# Streaming
# --------------------------------------------------------------------------- #


def _ndjson(event: dict[str, Any]) -> bytes:
    return (json.dumps(event, default=str) + "\n").encode("utf-8")


@router.post("/{project_id}/swarm-runs/{run_id}/stream")
async def stream(
    project_id: UUID,
    run_id: UUID,
    user_id: CurrentUser,
    conn: Connection = Depends(db_conn),
) -> StreamingResponse:
    _ensure_run_owner(conn, project_id, run_id, user_id)

    queue = broker.subscribe(run_id)

    async def gen():
        try:
            # Send initial snapshot so the UI has full state without an
            # extra REST call.
            snap = swarm_service.snapshot(run_id)
            yield _ndjson(snap)
            while True:
                event = await queue.get()
                if event is None:
                    break
                yield _ndjson(event)
        finally:
            broker.unsubscribe(run_id, queue)

    return StreamingResponse(gen(), media_type="application/x-ndjson")

File: src/frontend/src/lib/swarm.ts
// Swarm-run REST + NDJSON live event consumer.

export type SwarmType = 'aco' | 'pso' | 'abc' | 'firefly'
export type RunModel =
  | 'claude-opus-4-7'
  | 'claude-sonnet-4-6'
  | 'claude-haiku-4-5'
export type ThoughtLevel = 'minimal' | 'standard' | 'deep' | 'extreme'
export type RunStatus =
  | 'idle'
  | 'running'
  | 'paused'
  | 'completed'
  | 'failed'

export type SwarmRunConfig = {
  swarm_type: SwarmType
  model: RunModel
  num_agents: number
  thought_level: ThoughtLevel
  randomness: number
  pheromone_strength: number
}

export type SwarmRunSummary = {
  id: string
  project_id: string
  config: SwarmRunConfig
  status: RunStatus
  current_turn: number
  best_score: number | null
  created_at: string
  updated_at: string
}

export type SwarmRunDetail = SwarmRunSummary & {
  best_attempt_id: string | null
  best_system_text: string | null
  best_user_template: string | null
  error: string | null
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyJson = Record<string, any>

export type SwarmAttempt = {
  id: string
  turn: number
  agent_idx: number
  parent_attempt_id: string | null
  status: string
  score: number | null
  pheromone: number
  x: number | null
  y: number | null
  system_text: string | null
  user_template: string | null
  predicted_output: AnyJson | null
  error: string | null
}

export type SwarmAttemptSnapshot = {
  id: string
  turn: number
  agent_idx: number
  parent_attempt_id: string | null
  status: string
  system_preview: string
  user_preview: string
  score: number | null
  pheromone: number
  x: number | null
  y: number | null
  error: string | null
}

export type LayoutPoint = { id: string; x: number; y: number }

export type SwarmEvent =
  | {
      type: 'snapshot'
      run: SwarmRunDetail | (SwarmRunSummary & { error: string | null })
      attempts: SwarmAttemptSnapshot[]
      best: {
        id: string
        system_text: string
        user_template: string
        score: number
      } | null
      ts: number
    }
  | { type: 'turn_started'; turn: number }
  | {
      type: 'agent_started'
      turn: number
      agent_idx: number
      attempt_id: string
      parent_attempt_id: string | null
      parent_score: number | null
      parent_xy: { x: number; y: number } | null
    }
  | {
      type: 'agent_progress'
      turn: number
      agent_idx: number
      phase: 'sampling' | 'drafting' | 'scoring'
    }
  | {
      type: 'agent_drafted'
      turn: number
      agent_idx: number
      attempt_id: string
      system_preview: string
      user_preview: string
    }
  | {
      type: 'agent_scored'
      turn: number
      agent_idx: number
      attempt_id: string
      score: number
      pheromone: number
      latency_ms: number
      predicted_output: AnyJson
    }
  | {
      type: 'agent_failed'
      turn: number
      agent_idx: number
      attempt_id: string
      error: string
    }
  | { type: 'relayout'; turn: number; points: LayoutPoint[] }
  | {
      type: 'turn_complete'
      turn: number
      scores: number[]
      best_attempt_id: string | null
      best_score: number | null
    }
  | {
      type: 'best_updated'
      attempt_id: string
      score: number
      system_text: string
      user_template: string
    }
  | { type: 'paused' }
  | { type: 'completed' }
  | { type: 'error'; message: string }

type TokenGetter = () => Promise<string | null>

async function authedJson<T>(
  path: string,
  init: RequestInit,
  getToken: TokenGetter,
): Promise<T> {
  const token = await getToken()
  const headers = new Headers(init.headers)
  if (init.body !== undefined) headers.set('content-type', 'application/json')
  if (token) headers.set('authorization', `Bearer ${token}`)

  const res = await fetch(path, { ...init, headers })
  if (!res.ok) {
    let detail = res.statusText
    try {
      const body = await res.json()
      if (body?.detail) detail = body.detail
    } catch {
      /* ignore */
    }
    throw new Error(detail)
  }
  if (res.status === 204) return undefined as T
  return (await res.json()) as T
}

// --------------------------------------------------------------------------- //
// REST
// --------------------------------------------------------------------------- //

export function listSwarmRuns(
  projectId: string,
  getToken: TokenGetter,
): Promise<SwarmRunSummary[]> {
  return authedJson(
    `/api/projects/${projectId}/swarm-runs`,
    { method: 'GET' },
    getToken,
  )
}

export function createSwarmRun(
  projectId: string,
  config: SwarmRunConfig,
  getToken: TokenGetter,
): Promise<SwarmRunSummary> {
  return authedJson(
    `/api/projects/${projectId}/swarm-runs`,
    { method: 'POST', body: JSON.stringify(config) },
    getToken,
  )
}

export function getSwarmRun(
  projectId: string,
  runId: string,
  getToken: TokenGetter,
): Promise<SwarmRunDetail> {
  return authedJson(
    `/api/projects/${projectId}/swarm-runs/${runId}`,
    { method: 'GET' },
    getToken,
  )
}

export function startSwarmRun(
  projectId: string,
  runId: string,
  getToken: TokenGetter,
): Promise<{ started: boolean; status: RunStatus }> {
  return authedJson(
    `/api/projects/${projectId}/swarm-runs/${runId}/start`,
    { method: 'POST' },
    getToken,
  )
}

export function pauseSwarmRun(
  projectId: string,
  runId: string,
  getToken: TokenGetter,
): Promise<{ status: RunStatus }> {
  return authedJson(
    `/api/projects/${projectId}/swarm-runs/${runId}/pause`,
    { method: 'POST' },
    getToken,
  )
}

export function deleteSwarmRun(
  projectId: string,
  runId: string,
  getToken: TokenGetter,
): Promise<void> {
  return authedJson(
    `/api/projects/${projectId}/swarm-runs/${runId}`,
    { method: 'DELETE' },
    getToken,
  )
}

// --------------------------------------------------------------------------- //
// NDJSON stream
// --------------------------------------------------------------------------- //

async function* readNdjson(
  reader: ReadableStreamDefaultReader<Uint8Array>,
): AsyncGenerator<unknown> {
  const decoder = new TextDecoder()
  let buffer = ''
  while (true) {
    const { value, done } = await reader.read()
    if (done) break
    buffer += decoder.decode(value, { stream: true })
    let idx
    while ((idx = buffer.indexOf('\n')) !== -1) {
      const line = buffer.slice(0, idx).trim()
      buffer = buffer.slice(idx + 1)
      if (!line) continue
      try {
        yield JSON.parse(line)
      } catch {
        /* ignore malformed line */
      }
    }
  }
  buffer = buffer.trim()
  if (buffer) {
    try {
      yield JSON.parse(buffer)
    } catch {
      /* ignore */
    }
  }
}

export async function* streamSwarm(
  projectId: string,
  runId: string,
  getToken: TokenGetter,
  signal?: AbortSignal,
): AsyncGenerator<SwarmEvent> {
  const token = await getToken()
  const res = await fetch(
    `/api/projects/${projectId}/swarm-runs/${runId}/stream`,
    {
      method: 'POST',
      headers: {
        ...(token ? { authorization: `Bearer ${token}` } : {}),
      },
      signal,
    },
  )
  if (!res.ok || !res.body) {
    let detail = res.statusText
    try {
      const body = await res.json()
      if (body?.detail) detail = body.detail
    } catch {
      /* ignore */
    }
    throw new Error(detail)
  }
  const reader = res.body.getReader()
  for await (const ev of readNdjson(reader)) {
    yield ev as SwarmEvent
  }
}
